### PR TITLE
improved radix trie. Make use of radix trie in endpoint 'search user …

### PR DESCRIPTION
**Pull request by Christoph-Maximilian**
…by name'.

### CODE REVIEW

File: db_connector.py [hunk 1]
```diff
@@ -101,8 +100,21 @@ def get_all_users(self):
 
     def get_all_user_ids_pictures(self):
         logging.info("Retrieving all users ids and pictures")
-        users = list(r.db('loview').table('users').pluck('name', 'id', 'picture', 'age', 'gender').run(self.connection))
-        return users
+        latest_locations = pandas.DataFrame(list(r.db('loview').table('latest_locations').pluck("user_id", "lat", "lng", "time").run(self.connection)))
```
Comment (Christoph-Maximilian) [inline db_connector.py]: @muellerjonas I know we should do the outer join in rethinkdb - but I couldn`t get it work ... do you know how this can be done in the db?

Comment (muellerjonas) [inline db_connector.py]: I was just wondering... since we got static data now anyways why don't we just do some of the common queries we need to do at startup and cache their results?

File: run_server.py [hunk 2]
```diff
@@ -70,6 +70,16 @@ def get_user_details_and_locations_list_from_locations_dict(latest_locations_of_
...
 
+def create_filter_object(age_request_param, gender_request_param):
+    age = None
+    if not age_request_param is None:
+        age = int(age_request_param)
+    gender = None
+    if not gender_request_param is None:
+        gender = int(gender_request_param)
+        if not gender in [1, 2]:
+            gender = None
```
Comment (muellerjonas) [inline run_server.py]: logging.warning(...) in this case plz

File: run_server.py [hunk 3]
```diff
@@ -155,58 +165,29 @@ def user_lookup_by_id(user_id):  # this integrates data from multiple sources fo
...
-
-    user_ids_to_locations_map = {}
-    for user_id in user_ids:
-        latest_location_of_user, no_location = DB.get_latest_location_of_user(user_id["id"])
-        if no_location:
-            user_ids_to_locations_map[user_id["id"]] = {"server_status": ServerStatus.USER_HAS_NO_LOCATION, "lat": 0, "lng": 0}
+@app.route(REST.USER_LOOKUP_BY_NAME)
+def fast_user_lookup_by_name(user_name):
+    filter = create_filter_object(request.args.get('age'), request.args.get('gender'))
+    users_list = user_radix_trie.find_exact(key=user_name, filter=filter)
```
Comment (muellerjonas) [inline run_server.py]: Nice!

File: run_server.py [hunk 4]
```diff
@@ -375,15 +356,15 @@ def get_user_home_location_data(user_id):  # so we can reuse this function
...
-    user_autocompleter_rtrie.initialize(user_data)
+    user_radix_trie.initialize(user_data)
     logging.info("radix trie initialization done")
-    logging.info("size radix [B]:" + str(user_autocompleter_rtrie.get_size_bytes()))
+    logging.info("size radix [B]:" + str(user_radix_trie.get_size_bytes()))
     logging.info("size user data [B]:" + str(sys.getsizeof(user_data)))
-    logging.info("# nodes in trie:" + str(user_autocompleter_rtrie.get_number_nodes()))
-    app.run(debug=True, use_reloader=False)
+    logging.info("# nodes in trie:" + str(user_radix_trie.get_number_nodes()))
+    app.run(host='0.0.0.0', debug=True, use_reloader=False)
```
Comment (muellerjonas) [inline run_server.py]: Is this addition necessary?

File: db_connector.py (hunk 1 above)
Comment (muellerjonas) [reply db_connector.py]: No I don't
If you couldn't get it to work by checking the docs then pandas should be fine if it performs well

Comment (Christoph-Maximilian) [reply db_connector.py]: yes this would be a nice feature
